datasource/src/main/scala/quasar/physical/s3/RequestSigning.scala

```diff
@@ -0,0 +1,201 @@
+    signedHeaders(req.uri.path, req.method, req.params, req.headers, req.body)
+
+  def signedHeaders[F[_]: Sync](path: Uri.Path, method: Method, queryParams: Map[String, String], headers: Headers, payload: Stream[F, Byte]): F[Headers] = {
+
+    val now: LocalDateTime = clock
+    val credentialsNow = credentials
+
+    val extraSecurityHeaders: Headers =
+      Headers(credentialsNow.sessionToken.toList map xAmzSecurityTokenHeader)
+
+    val extraDateHeaders: Headers =
+      if (!headers.iterator.exists(_.name == Date.name)) Headers(xAmzDateHeader(now)) else Headers()
+
+    val signedHeaders = headers ++ extraDateHeaders ++ extraSecurityHeaders
+
+    val signedHeaderKeys = signedHeaders.toList.map(_.name.value.toLowerCase).sorted.mkString(";")
+
+    val sha256Payload: F[String] =
+      payloadSigning match {
+        case PayloadSigning.Unsigned => "UNSIGNED-PAYLOAD".pure[F]
+        case PayloadSigning.Signed   => sha256(payload) map base16
+      }
+
+    sha256Payload map { payloadHash =>
+      val canonicalRequest =
+        List(
+          method.toString,
+          path.toString,
+          renderCanonicalQueryString(queryParams),
+          renderCanonicalHeaders(signedHeaders),
+          signedHeaderKeys,
+          payloadHash,
+        ).mkString("\n")
+
+      val credentialScope: String =
+        List(
+          now.format(DateTimeFormatter.BASIC_ISO_DATE),
+          region.name,
+          service.name + "/aws4_request"
+        ).mkString("/")
+
+      val stringToSign =
+        List(
+          "AWS4-HMAC-SHA256",
+          now.format(DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'")),
+          credentialScope,
+          base16(sha256(canonicalRequest.getBytes(StandardCharsets.UTF_8)))
+        ).mkString("\n")
+
+      val signature = sign(stringToSign, now, credentialsNow, region, service)
+
+      val authorizationHeaderValue =
+        "AWS4-HMAC-SHA256 Credential=" +
+          credentialsNow.accessKey.value + "/" + credentialScope +
+          ", SignedHeaders=" + signedHeaderKeys +
+          ", Signature=" + signature
+
+      val payloadChecksumHeader: Header =
+        payloadSigning match {
+          case PayloadSigning.Unsigned => xAmzContentSha256("UNSIGNED-PAYLOAD")
+          case PayloadSigning.Signed   => xAmzContentSha256(payloadHash)
+        }
+
+      signedHeaders.put(
+        Header(Authorization.name.value, authorizationHeaderValue),
+        payloadChecksumHeader,
+      )
+    }
+  }
+}
+
+@SuppressWarnings(Array("org.wartremover.warts.DefaultArguments"))
+final case class Credentials(
+    accessKey:    AccessKey,
+    secretKey:    SecretKey,
+    sessionToken: Option[String] = None)
+
+abstract class ServiceName(val name: String)
+
+object ServiceName {
+  object S3 extends ServiceName("s3")
+}
+
+sealed trait PayloadSigning
+
+object PayloadSigning {
+
+  /** Payload is signed. Note that it will cause the payload stream to be consumed twice. */
```

Comment:
Why is it the case that the stream will be consumed twice?

Reply:
For this PR, I don't think this has any effect at all since none of our requests have payloads (we only use `HEAD` and `GET`). AFAIK, it needs to consume it twice since it needs to consume it once to hash it and once again to submit it.